rust/src/database/verkle/variants/simple/node.rs

```diff
@@ -0,0 +1,583 @@
+    Empty,
+    Inner(InnerNode),
+    Leaf(LeafNode),
+}
+
+#[cfg(test)]
+impl Node {
+    /// Returns the value associated with the given key, or the default value if
+    /// the key does not exist.
+    pub fn get(&self, key: &Key, depth: u8) -> Value {
+        match self {
+            Node::Empty => Value::default(),
+            Node::Inner(inner) => inner.get(key, depth),
+            Node::Leaf(leaf) => leaf.get(key),
+        }
+    }
+
+    /// Sets the value for the given key.
+    /// Consumes the node and returns an updated version.
+    pub fn set(self, key: &Key, depth: u8, value: &Value) -> Node {
+        match self {
+            Node::Empty => {
+                if depth == 0 {
+                    // While conceptually it would suffice to create a leaf node here,
+                    // Geth always creates an inner node (and we want to stay compatible).
+                    let inner = InnerNode::new();
+                    inner.set(key, depth, value)
+                } else {
+                    let leaf = LeafNode::new(key);
+                    leaf.set(key, depth, value)
+                }
+            }
+            Node::Inner(inner) => inner.set(key, depth, value),
+            Node::Leaf(leaf) => leaf.set(key, depth, value),
+        }
+    }
+
+    /// Computes and returns the commitment of this node.
+    ///
+    /// If the commitment is already up to date, it is returned without recomputation.
+    pub fn commit(&mut self) -> Commitment {
+        match self {
+            Node::Empty => Commitment::default(),
+            Node::Inner(inner) => inner.commit(),
+            Node::Leaf(leaf) => leaf.commit(),
+        }
+    }
+
+    /// Returns the current commitment of this node without recomputing it.
+    fn get_commitment(&self) -> Commitment {
+        match self {
+            Node::Empty => Commitment::default(),
+            Node::Inner(inner) => inner.commitment,
+            Node::Leaf(leaf) => leaf.commitment,
+        }
+    }
+
+    /// Returns true if the commitment of this node is dirty and needs to be recomputed.
+    fn commitment_is_dirty(&self) -> bool {
+        match self {
+            Node::Empty => false,
+            Node::Inner(inner) => inner.commitment_dirty,
+            Node::Leaf(leaf) => leaf.commitment_dirty,
+        }
+    }
+}
+
+/// An inner node in the simple in-memory Verkle trie, containing up to 256 children.
+#[cfg_attr(not(test), expect(unused))]
+#[derive(Debug)]
+pub struct InnerNode {
+    children: Vec<Node>,
+    commitment: Commitment,
+    commitment_dirty: bool,
+}
+
+#[cfg(test)]
+impl InnerNode {
+    /// Creates a new inner node without any children.
+    fn new() -> Self {
+        let mut children = Vec::with_capacity(256);
+        for _ in 0..256 {
+            children.push(Node::Empty);
+        }
+        InnerNode {
+            children,
+            commitment: Commitment::default(),
+            commitment_dirty: true,
+        }
+    }
```

Comment:
Sorry for my urge of having to encode everything in the type system, but if we keep the allocation here then please use a boxed array instead.
```suggestion
pub struct InnerNode {
    children: Box<[Node; 256]>,
    commitment: Commitment,
    commitment_dirty: bool,
}

#[cfg(test)]
impl InnerNode {
    /// Creates a new inner node without any children.
    fn new() -> Self {
        let children = Box::new(std::array::from_fn(|_| Node::Empty));
        InnerNode {
            children,
            commitment: Commitment::default(),
            commitment_dirty: true,
        }
    }
```